packages/openid4vc-client/src/OpenId4VcClientService.ts

```diff
@@ -1,10 +1,164 @@
+    return {
+      signCallback: this.signCallback(agentContext),
+    }
+  }
+
+  public async preAuthorized(agentContext: AgentContext, options: PreAuthorizedOptions): Promise<W3cCredentialRecord> {
+    this.logger.debug('Running pre-authorized flow with options', options)
+
+    const client = await OpenID4VCIClient.initiateFromURI({
+      issuanceInitiationURI: options.issuerUri,
+      flowType: AuthzFlowType.PRE_AUTHORIZED_CODE_FLOW,
+      kid: options.kid,
+      alg: Alg.EdDSA,
+      // The clientId is set to a dummy value for now as we don't have
+      // one in the pre-auth flow, but it's currently required by OpenID4VCIClient.
+      clientId: 'test-clientId',
+    })
+
+    // The clientId is set to a dummy value for now as we don't have
+    // one in the pre-auth flow, but it's currently required by OpenID4VCIClient.
+    const accessToken = await client.acquireAccessToken({ clientId: 'test-clientId' })
+
+    this.logger.info('Fetched server accessToken', accessToken)
+
+    // We currently need the ts-ignore because the type
+    // inside of OpenID4VCIClient needs to be updated.
+    // @ts-ignore
+    if (!accessToken.scope) {
+      throw new AriesFrameworkError(
+        "Access token response doesn't contain a scope. Only scoped issuer URIs are supported at this time."
+      )
+    }
+
+    const serverMetadata = await client.retrieveServerMetadata()
+
+    this.logger.info('Fetched server metadata', {
+      issuer: serverMetadata.issuer,
+      credentialEndpoint: serverMetadata.credential_endpoint,
+      tokenEndpoint: serverMetadata.token_endpoint,
+    })
+
+    this.logger.debug('Full server metadata', serverMetadata)
+
+    // proof of possession
+    const callbacks = this.getSignCallback(agentContext)
+
+    const proofInput = await ProofOfPossessionBuilder.fromAccessTokenResponse({
+      accessTokenResponse: accessToken,
+      callbacks: callbacks,
+    })
+      .withEndpointMetadata(serverMetadata)
+      // The clientId is set to a dummy value for now as we don't have
+      // one in the pre-auth flow, but it's currently required by OpenID4VCIClient.
+      .withClientId('test-clientId')
+      .withKid(options.kid)
+      .build()
+
+    this.logger.debug('Generated JWS', proofInput)
+
+    const credentialRequestClient = CredentialRequestClientBuilder.fromIssuanceInitiationURI({
+      uri: options.issuerUri,
+      metadata: serverMetadata,
+    })
+      .withTokenFromResponse(accessToken)
+      .build()
+
+    const credentialResponse = await credentialRequestClient.acquireCredentialsUsingProof({
+      proofInput,
+      // @ts-ignore
+      credentialType: accessToken.scope,
+      format: 'ldp_vc', // Allows us to override the format
+    })
+
+    this.logger.debug('Credential request response', credentialResponse)
+
+    if (!credentialResponse.successBody) {
+      throw new AriesFrameworkError('Did not receive a successful credential response')
+    }
+
```

Comment:
I think it would be good to verify the received credential before storing, just so we know the signatuer is valid and we'll be able to use i in the future. We also do this in the `JsonLdCredentialFormatService`.:

```
// verify signatures of the credential
    const result = await w3cCredentialService.verifyCredential(agentContext, { credential })
    if (result && !result.verified) {
      throw new AriesFrameworkError(`Failed to validate credential, error = ${result.error}`)
    }
```

Reply:
Not for now, just an idea. We should probably also start exposing an `W3cCredentialApi` so people can use the public api to find/store/validate w3c credentials. Then we also wouldn't have to depend on internal Apis here.